server/proto/merger.go

```diff
@@ -0,0 +1,228 @@
+				lastIndex := md.Messages().Len() - 1
+				if md.Messages().Get(lastIndex) == prevMsgDesc {
+					parent = parents[len(parents)-1]
+					parents[len(parents)-1] = nil
+				}
+			}
+		}
+		return true
+	})
+}
+
+// Get FileDescriptor from a registry by path and convert it to FileDescriptorProto
+func getFileDescriptorProto(registry *protoregistry.Files, path string) *descriptorpb.FileDescriptorProto {
+	fileDescriptor, err := registry.FindFileByPath(path)
+	if err != nil {
+		return nil
+	}
+	return protodesc.ToFileDescriptorProto(fileDescriptor)
+}
+
+// add new files from current registry to previous registry, store them in a list
+func addNewFiles(
+	currRegistry *protoregistry.Files,
+	prevRegistry *protoregistry.Files,
+	mergedFiles []*descriptorpb.FileDescriptorProto,
+) []*descriptorpb.FileDescriptorProto {
+	currRegistry.RangeFiles(func(currFileDesc protoreflect.FileDescriptor) bool {
+		currFile := protodesc.ToFileDescriptorProto(currFileDesc)
+		if getFileDescriptorProto(prevRegistry, currFileDesc.Path()) == nil {
+			mergedFiles = append(mergedFiles, currFile)
+		}
+		return true
+	})
+	return mergedFiles
+}
+
+// add new message from currMsgs to prevMsgs
+func addMessage(prevMsgs *[]*descriptorpb.DescriptorProto, currMsgs *[]*descriptorpb.DescriptorProto) {
+	for _, curr := range *currMsgs {
+		exist := false
+		for _, prev := range *prevMsgs {
+			if prev.GetName() == curr.GetName() {
+				exist = true
+				break
+			}
+		}
+		if !exist {
+			*prevMsgs = append(*prevMsgs, curr)
+		}
+	}
+}
+
+// Append all new message in current file to previous file.
+func addNewMessage(prevFile *descriptorpb.FileDescriptorProto, currFile *descriptorpb.FileDescriptorProto) {
+	addMessage(&prevFile.MessageType, &currFile.MessageType)
+}
+
+// Append all new nested message in current message to previous message.
+func addNewNestedMessage(prevParent *descriptorpb.DescriptorProto, currParent *descriptorpb.DescriptorProto) {
+	addMessage(&prevParent.NestedType, &currParent.NestedType)
+}
+
+// Add all new import statements in current file to previous file.
+func addNewDependencies(prevFile *descriptorpb.FileDescriptorProto, currFile *descriptorpb.FileDescriptorProto) {
+	for _, currDep := range currFile.GetDependency() {
+		isNew := true
+		for _, prevDep := range prevFile.GetDependency() {
+			if currDep == prevDep {
+				isNew = false
+				break
+			}
+		}
+		if isNew {
+			prevFile.Dependency = append(prevFile.Dependency, currDep)
+		}
+	}
+}
+
+// Set deprecated options to true for each removed field in message.
+func deprecateRemovedFields(prev *descriptorpb.DescriptorProto, curr *descriptorpb.DescriptorProto) {
+	for _, prevField := range prev.Field {
+		removed := true
+		for _, currField := range curr.Field {
+			if prevField.GetName() == currField.GetName() {
+				removed = false
+				break
+			}
+		}
+		if removed {
+			deprecated := true
+			prevField.Options = &descriptorpb.FieldOptions{Deprecated: &deprecated}
```

Comment:
retain existing options data if exists then add deprecated field. This might overwrite any existing field options.